README.md
# aprsWsprBridge
simple python script(s) to collect wspr spots from wsprnet.org and upload a position report to aprs.fi


changelog.txt
#==============================================================================================================#
#
#  2020-05-06 - changed time.mktime to calendar.timegm in line:
#                wsprSpotTimeStamp = int( calendar.timegm( wsprSpotDateTime.timetuple() ) )
#             - this was a bug reported by Bob ZL1RS which identified a problem with local time vs UTC time 
#               conversion.
#
#==============================================================================================================#


getWspr.py
#!/usr/bin/python
#==============================================================================================================#
#                                                                                                              #
# getWspr                                                                                                      #
#                                                                                                              #
# Copyright (C) 2019 Graham Collins ve3gtc                                                                     #
#                                                                                                              #
# This program is free software; you can redistribute it and/or modify                                         #
# it under the terms of the GNU General Public License as published by                                         #
# the Free Software Foundation; either version 2 of the License, or                                            #
# (at your option) any later version.                                                                          #
#                                                                                                              #
# This program is distributed in the hope that it will be useful,                                              #
# but WITHOUT ANY WARRANTY; without even the implied warranty of                                               #
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the                                                #
# GNU General Public License for more details.                                                                 #
#                                                                                                              #
# You should have received a copy of the GNU General Public License along                                      #
# with this program; if not, write to the Free Software Foundation, Inc.,                                      #
# 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.                                                  #
#                                                                                                              #
#==============================================================================================================#
#
# if not already installled, use pip to install the following
#
#    pip install beautifulsoup
#    pip install bs4
#    pip install lxml
#    pip install pandas
#    pip install requests
#
#==============================================================================================================#

import sys
import traceback
import string
import time
from bs4 import BeautifulSoup
import requests
import pandas as pd


"""
Contains a function for getting wspr spots from wsprnet.org
"""

def getWspr( gwStrCallsign ):
	"""
	Takes a CALLSIGN and gets WSPR spots for that callsign from wsprnet.org
	"""

	with open("log.txt", "a") as log_file:
		log_file.write( time.ctime() + " : getWspr start\n" )

	url = 'http://wsprnet.org/olddb?mode=html&band=all&limit=50&findcall=' + gwStrCallsign + '&findreporter=&sort=date'


	try:
		r = requests.get(url)
		r.raise_for_status()

	except requests.exceptions.HTTPError as errh:

		with open( 'log.txt', 'a' ) as log_file:
			log_file.write( time.ctime() + ' : http error %s \n' % errh)
			log_file.close()

		return 1

	except requests.exceptions.ConnectionError as errc:

		with open( 'log.txt', 'a' ) as log_file:
			log_file.write( time.ctime() + ' : error connecting %s \n' % errc)
			log_file.close()

		return 1

	except requests.exceptions.Timeout as errt:

		with open( 'log.txt', 'a' ) as log_file:
			log_file.write( time.ctime() + ' : connection timeout %s \n' % errt)
			log_file.close()

		return 1

	except requests.exceptions.RequestException as err:

		with open( 'log.txt', 'a' ) as log_file:
			log_file.write( time.ctime() + ' : some other connection error %s \n' % err )
			log_file.close()

		return 1

	data = r.text

	pd.set_option('display.expand_frame_repr', False)

	soup = BeautifulSoup(data, 'lxml')

	table = soup.find_all('table')[2]            # want third table

	rows = table.find_all('tr')[2:]      # skip first two rows

	data = {
		'dateTime' : [],
		'Call' : [],
		'Frequency' : [],
		'SNR' : [],
		'drift' : [],
		'Grid' : [],
		'powerDBM' : [],
		'powerWatts' : [],
		'reporter' : [],
		'reporterGrid' : [],
		'distanceKM' : [],
		'distanceMI' : []
		}

	try:
		# extract values and remove unicode &nbsp;
		for row in rows:
			cols = row.find_all('td')
			data['dateTime'].append( cols[0].get_text().replace(u'\xa0', "") )
			data['Call'].append( cols[1].get_text().replace(u'\xa0', "") )
			data['Frequency'].append( cols[2].get_text().replace(u'\xa0', "") )
			data['SNR'].append( cols[3].get_text().replace(u'\xa0', "") )
			data['drift'].append( cols[4].get_text().replace(u'\xa0', "") )
			data['Grid'].append( cols[5].get_text().replace(u'\xa0', "") )
			data['powerDBM'].append( cols[6].get_text().replace(u'\xa0', "") )
			data['powerWatts'].append( cols[7].get_text().replace(u'\xa0', "") )
			data['reporter'].append( cols[8].get_text().replace(u'\xa0', "") )
			data['reporterGrid'].append( cols[9].get_text().replace(u'\xa0', "") )
			data['distanceKM'].append( cols[10].get_text().replace(u'\xa0', "") )
			data['distanceMI'].append( cols[11].get_text().replace(u'\xa0', "") )

	except IndexError:

		with open("log.txt", "a") as log_file:
			log_file.write( time.ctime() + " : no wspr spots for %s\n" % gwStrCallsign )
			return 1

	spots = pd.DataFrame( data )

	# will return only the first spot from the list which is the latest spot reported by wsprnet.org
	Spot = spots['dateTime'].values[0]
	Spot = Spot + " " + spots['Call'].values[0]
	Spot = Spot + " " + spots['Frequency'].values[0]
	Spot = Spot + " " + spots['SNR'].values[0]
	Spot = Spot + " " + spots['drift'].values[0]
	Spot = Spot + " " + spots['Grid'].values[0]
	Spot = Spot + " " + spots['powerDBM'].values[0]
	#Spot = Spot + " " + spots['powerWatts'].values[0]      # commented out
	Spot = Spot + " " + spots['reporter'].values[0]
	Spot = Spot + " " + spots['reporterGrid'].values[0]
	#Spot = Spot + " " + spots['distanceKM'].values[0]      # commented out
	Spot = Spot + " " + spots['distanceMI'].values[0]

	with open("log.txt", "a") as log_file:
		log_file.write(time.ctime() + " : wspr spots collected\n")


	return Spot


aprsUtilities.py
#!/usr/bin/python
#==============================================================================================================#
#                                                                                                              #
# aprsUtilies - Python library of utilities for use with posting spots to APRS                                 #
#                                                                                                              #
# Copyright (C) 2019 Graham Collins ve3gtc                                                                     #
#                                                                                                              #
# This program is free software; you can redistribute it and/or modify                                         #
# it under the terms of the GNU General Public License as published by                                         #
# the Free Software Foundation; either version 2 of the License, or                                            #
# (at your option) any later version.                                                                          #
#                                                                                                              #
#  http://www.gnu.org/licenses/gpl-3.0.html                                                                    #
#                                                                                                              #
# This program is distributed in the hope that it will be useful,                                              #
# but WITHOUT ANY WARRANTY; without even the implied warranty of                                               #
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the                                                #
# GNU General Public License for more details.                                                                 #
#                                                                                                              #
# You should have received a copy of the GNU General Public License along                                      #
# with this program; if not, write to the Free Software Foundation, Inc.,                                      #
# 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.                                                  #
#                                                                                                              #
#==============================================================================================================#
#
# if not already installled, use pip to install the following
#
#    pip install beautifulsoup
#    pip install bs4
#    pip install lxml
#    pip install pandas
#    pip install requests
#
#==============================================================================================================#

import sys
import traceback
import string
from socket import *
import re                        #regex
import datetime
import time
import requests
import json

#==============================================================================================================#
#                                                                                                              #
# aprslib - Python library for working with APRS                                                               #
# Copyright (C) 2013-2014 Rossen Georgiev                                                                      #
#                                                                                                              #
# https://github.com/rossengeorgiev/aprs-python/blob/master/aprslib/passcode.py                                #
#                                                                                                              #
# This program is free software; you can redistribute it and/or modify                                         #
# it under the terms of the GNU General Public License as published by                                         #
# the Free Software Foundation; either version 2 of the License, or                                            #
# (at your option) any later version.                                                                          #
#                                                                                                              #
# This program is distributed in the hope that it will be useful,                                              #
# but WITHOUT ANY WARRANTY; without even the implied warranty of                                               #
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the                                                #
# GNU General Public License for more details.                                                                 #
#                                                                                                              #
# You should have received a copy of the GNU General Public License along                                      #
# with this program; if not, write to the Free Software Foundation, Inc.,                                      #
# 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.                                                  #
#                                                                                                              #
#==============================================================================================================#
"""
Contains a function for generating passcode from callsign
"""

def passCode(pcStrCallsign):
    """
    Takes a CALLSIGN and returns passcode
    """
    assert isinstance(pcStrCallsign, str)

    pcStrCallsign = pcStrCallsign.split('-')[0].upper()

    code = 0x73e2
    for i, char in enumerate(pcStrCallsign):
        code ^= ord(char) << (8 if not i % 2 else 0)

    return code & 0x7fff


#==============================================================================================================#
#                                                                                                              #
# - AprsSendPacket(aspStrCallSign, aspStrAprsMessage)                                                 #
#                                                                                                              #
# - aspStrCallSign is the callsign of WHO is posting the spot, NOT the callsign of the ballaoon                #
# - the callsign of balloon is in aspStrAprsMessage                                                            #
#                                                                                                              #
# - sends data to aprs.fi                                                                                      #
#                                                                                                              #
# - http://www.aprs-is.net                                                                                     #
#                                                                                                              #
#==============================================================================================================#
"""
function to upload a data packet to aprs.fi netowrk
"""
#def AprsSendPacket(aspStrCallSign, apsStrAprsMessage):

def AprsSendPacket(aspStrUserCallsign, aspStrBalloonCAllSign, aspStrLatitude, aspStrLongitude, aspStrIcon, aspStrData):

	serverHost = 'second.aprs.net'
	serverPort = 10152

	#==============================================================================================================#
	#                                                                                                              #
	# Very simple RegEx to determine is CallSign string is a valid amateur call sign:                              #
	#                                                                                                              #
	# Determines if the callsign conforms to amateur radio international rules (#A#, #AA#, # A#, A##, or AA#).     #
	#                                                                                                              #
	# Callsign length must be between 3 and 6 characters.                                                          #
	#                                                                                                              #
	# @return true if callsign is compliant                                                                        #
	#                                                                                                              #
	# only continue to upload to APRS if provided callsign is compliant                                            #
	#                                                                                                              #
	#==============================================================================================================#

	if ( re.search('^(?:(?:[1-9][A-Z][A-Z]?)|(?:[A-Z][2-9A-Z]?))[0-9][A-Z]{1,3}$', aspStrUserCallsign) ):

		# Pass - compliant callsign, make log entry
		with open("log.txt", "a") as log_file:
			log_file.write(time.ctime() + ' : ' + aspStrUserCallsign + ' is a compliant callsign \n' )

		intPassCode = passCode(aspStrUserCallsign)

		# create socket & connect to server
		clientSocket = socket(AF_INET, SOCK_STREAM)
		clientSocket.connect((serverHost, serverPort))

		# Aprs log in 'user CallSign pass ' + intPassCode + ' vers "aprsBallloon 0.1" \n'
		strAprsLogIn = 'user ' + aspStrUserCallsign + ' pass ' + str(intPassCode) + ' vers "aprsBalloon 0.1" \n'

		clientSocket.sendto(strAprsLogIn.encode(),(serverHost, serverPort))

		# AprsSendPacket(aspStrUserCallsign, aspStrBalloonCAllSign, aspStrLatitude, aspStrLongitude, aspStrIcon, aspStrData):

		aspStrAprsMessage = aspStrBalloonCAllSign + '>APRS,TCIP*:=' + aspStrLatitude + '/' + aspStrLongitude + aspStrIcon + aspStrData + '\n'

		# log packet to be sent
		with open("log.txt", "a") as log_file:
			log_file.write(time.ctime() + ' : ' + aspStrAprsMessage )

		# send aspStrAprsMessage
		clientSocket.sendto(aspStrAprsMessage.encode(),(serverHost, serverPort))

		# close socket -- must be closed to avoidbuffer overflow
		time.sleep(5)                                                        # 5 sec. delay
		clientSocket.shutdown(0)
		clientSocket.close()

		# log packet sent event
		with open("log.txt", "a") as log_file:
			log_file.write(time.ctime() + " : Packet sent at: "+ time.ctime() + '\n' )

		return None


#==============================================================================================================#
#                                                                                                              #
# latLongToGrid(Latitude, Longitude)                                                                           #
#                                                                                                              #
# - converts decimal latitude and Longitude to six character Maidenhead Grid Square                            #
#                                                                                                              #
# - returns Gridsquare as string                                                                               #
#                                                                                                              #
# - using convention of North + South - West - and East +                                                      #
#                                                                                                              #
#==============================================================================================================#
def latLongToGrid(decLatitude, decLongitude):

	upper = 'ABCDEFGHIJKLMNOPQRSTUVWX'
	lower = 'abcdefghijklmnopqrstuvwx'

	llgLatitude  = decLatitude  + 90
	llgLongitude = decLongitude + 180

	gridSquare = upper[int(llgLongitude/20)]
	gridSquare = "%s%s" % (gridSquare, upper[int(llgLatitude/10)])

	gridSquare = "%s%s" % (gridSquare, str(int((llgLongitude/2)%10)))
	gridSquare = "%s%s" % (gridSquare, str(int(llgLatitude%10)))

	gridSquare = "%s%s" % (gridSquare, lower[int(((llgLongitude - int(llgLongitude)) * 60)/5)])
	gridSquare = "%s%s" % (gridSquare, lower[int(((llgLatitude - int(llgLatitude)) * 60)/2.5)])

	return gridSquare

#==============================================================================================================#
#                                                                                                              #
# gridToLatLong(GridSquare)                                                                                    #
#                                                                                                              #
# - converts 4, or 6, charcater Maidenhead Grid Square to decimal Latitude and Longitude                       #
#                                                                                                              #
# - returns Latitude/Longitude as string                                                                       #
#                                                                                                              #
# - using convention of North + South - West - and East +                                                      #
#                                                                                                              #
#==============================================================================================================#
def gridToLatLong(gllGridSquare):

	if not isValidGridSquare(gllGridSquare):
		return "errorInvalidGrid"

	gllGridSquare = gllGridSquare.upper()

	gllLongitude = -180 + ( ((ord(gllGridSquare[0]) - ord('A') ) * 20) + (int(gllGridSquare[2]) * 2 ) )
	gllLatitude  =  -90 + ( ((ord(gllGridSquare[1]) - ord('A') ) * 10) + (int(gllGridSquare[3]) ) )

	if len(gllGridSquare) > 4 :
		gllLongitude = gllLongitude + ( float( (ord(gllGridSquare[4]) - ord('A') ) * 5)   / 60 )
		gllLatitude  =  gllLatitude + ( float( (ord(gllGridSquare[5]) - ord('A') ) * 2.5) / 60 )

	if len(gllGridSquare) == 8 :
		gllLongitude = gllLongitude + ( ( float(int(gllGridSquare[6])) * .5  ) / 60 )
		gllLatitude  = gllLatitude  + ( ( float(int(gllGridSquare[7])) * .25 ) / 60 )

	return "%s %s" % (gllLatitude,gllLongitude)

#==============================================================================================================#
#                                                                                                              #
# isValidGridSquare(GridSqure)                                                                                 #
#                                                                                                              #
# - determines if GridSquare is a valid 4, 6, or 8 charcater Maidenhead Grid Square                            #
#                                                                                                              #
# - returns TRUE or FALSE                                                                                      #
#                                                                                                              #
# - Grid Square pattern is one of 4, 6 or 8 characters and numbers in one of the following                     #
#   patterns:                                                                                                  #
#              [A-R][A-R][0-9][0-9]                                                                            #
#              [A-R][A-R][0-9][0-9][a-x][a-x]                                                                  #
#              [A-R][A-R][0-9][0-9][a-x][a-x][0-9][0-9]                                                        #
#                                                                                                              #
# - convention is that the first pair of letters is upper case and the second set of                           #
#   letters is lower case but in reality Maidenhead Grid Squares are case insensitive                          #
#                                                                                                              #
#==============================================================================================================#
def isValidGridSquare(GridSquare):

	GridSquare = GridSquare.upper()

	if len(GridSquare) == 4 :
		if re.match( '^[A-R][A-R][0-9][0-9]$', GridSquare):
			return True
		else:
			return False

	elif len(GridSquare) == 6 :
		if re.match( '^[A-R][A-R][0-9][0-9][A-X][A-X]$', GridSquare):
			return True
		else:
			return False

	elif len(GridSquare) == 8 :
		if re.match( '^[A-R][A-R][0-9][0-9][A-X][A-X][0-9][0-9]$', GridSquare):
			return True
		else:
			return False

	else:
		return False

#==============================================================================================================#
#                                                                                                              #
# toDegreesMinutes(value,type)                                                                                 #
#                                                                                                              #
# - converts decimal degress to degrees minutes                                                                #
#                                                                                                              #
#==============================================================================================================#
def toDegreesMinutes(value,type):
	"""
        Converts a DD.dddd value into DDMM.mm Notation.

        Pass value as double
        type = {Latitude or Longitude} as string

        returns a string as DDMM.mmDirection

        modified from http://anothergisblog.blogspot.ca/2011/11/convert-decimal-degree-to-degrees.html
	"""
	try:
		value = float(value)
	except:
		#print '\nERROR: Could not convert %s to float.' %(type(value))

		with open("log.txt", "a") as log_file:
			log_file.write(time.ctime() + ' : could not convert ' + str(value) + ' to float\n')

		return 0

	degrees = int(value)
	minutes = abs( (value - int(value) ) * 60.0)

	quadrant = ""
	if type == "Longitude":
		if degrees <= 0:
			quadrant = "W"
		elif degrees > 0:
			quadrant = "E"
		else:
			quadrant = ""
	elif type == "Latitude":
		if degrees < 0:
			quadrant = "S"
		elif degrees >= 0:
			quadrant = "N"
		else:
			quadrant = ""

	if type == "Longitude":
		degreesMinutes = str('{:0>3d}'.format(abs(degrees))) + str('{:05.2f}'.format(minutes)) + "" + quadrant

	elif type == "Latitude":
		degreesMinutes = str('{:0>2d}'.format(abs(degrees))) + str('{:05.2f}'.format(minutes)) + "" + quadrant

	return degreesMinutes


aprsWsprBridge.py
#!/usr/bin/python
#==============================================================================================================#
#                                                                                                              #
# aprsWsprBridge.py - simple python script to collect WSPR spots and upload a position to APRS                 #
#                                                                                                              #
# Copyright (C) 2019 Graham Collins ve3gtc                                                                     #
#                                                                                                              #
# This program is free software; you can redistribute it and/or modify                                         #
# it under the terms of the GNU General Public License as published by                                         #
# the Free Software Foundation; either version 2 of the License, or                                            #
# (at your option) any later version.                                                                          #
#                                                                                                              #
#  http://www.gnu.org/licenses/gpl-3.0.html                                                                    #
#                                                                                                              #
# This program is distributed in the hope that it will be useful,                                              #
# but WITHOUT ANY WARRANTY; without even the implied warranty of                                               #
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the                                                #
# GNU General Public License for more details.                                                                 #
#                                                                                                              #
# You should have received a copy of the GNU General Public License along                                      #
# with this program; if not, write to the Free Software Foundation, Inc.,                                      #
# 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.                                                  #
#                                                                                                              #
#==============================================================================================================#
#
# if not already installled, use pip to install the following
#
#    pip install beautifulsoup
#    pip install bs4
#    pip install lxml
#    pip install pandas
#    pip install requests
#
#==============================================================================================================#
#
#  2020-05-06 - changed time.mktime to calendar.timegm in line:
#                wsprSpotTimeStamp = int( calendar.timegm( wsprSpotDateTime.timetuple() ) )
#             - this was a bug reported by Bob ZL1RS which identified a problem with local time vs UTC time 
#               conversion.
#
#==============================================================================================================#

import sys
import traceback
import string
import datetime
import time
import calendar

from aprsUtilities import *
from getWspr import *
from getLastAprsPositionTime import *

#------------------------ starts here -----------------------------#

with open( "log.txt", "a" ) as log_file:
    log_file.write( time.ctime() + " :#----------------------------------------------------------------------------------------------------------------------------------\n" )

if sys.version_info[0] > 3 :
	raise NameError( 'Tested on only Python ver 3 ')

else :
	with open( "log.txt", "a" ) as log_file:
		log_file.write( time.ctime() + " : Cleared to continue\n" )

strUserCallSign = 'your callsign - the one who will be uploading the spot to APRS'
aprsApiKey = 'your aprs api key see: https://aprs.fi/page/api you will need to create and account and you will find your key here: https://aprs.fi/account/'

strWsprCallSign = 'callsign of the wspr target'

strBalloonCallSign = 'callsign you want your balloon to appear as on APRS i.e. W1XXX-12'

strIcon = 'O'                                        # balloon

wsprSpot = getWspr( strWsprCallSign )

# 2020-01-15 22:10 VE3GHM 14.097184 -26 0 FN25ig +10 KD6RF EM22 1375

if wsprSpot == 1 :                                   # some sort of exception occured, check log

	with open( 'log.txt', 'a' ) as log_file:
		log_file.write( time.ctime() + ' : connection error or no wspr spots encountered \n' )
		log_file.write( time.ctime() + " :#----------------------------------------------------------------------------------------------------------------------------------\n" )
		log_file.close()
		sys.exit( 0 )

lastAprsPositionTimeStamp = int( getLastAprsPositionTime ( strBalloonCallSign, aprsApiKey ) )

# print ( lastAprsPositionTimeStamp )

wsprSpotDetails = wsprSpot.split()

wsprDateTime = wsprSpotDetails[0] + ' ' + wsprSpotDetails[1]

wsprSpotDateTime = datetime.datetime.strptime( wsprDateTime, '%Y-%m-%d %H:%M' )

wsprSpotTimeStamp = int( calendar.timegm( wsprSpotDateTime.timetuple() ) )

with open( "log.txt", "a" ) as log_file:
	log_file.write( time.ctime() + " : last APRS postion timestamp %s\n" % lastAprsPositionTimeStamp )

with open( "log.txt", "a" ) as log_file:
	log_file.write( time.ctime() + " : last WSPR postion timestamp %s\n" % wsprSpotTimeStamp )

if wsprSpotTimeStamp > lastAprsPositionTimeStamp + 15 :

	strGrid = wsprSpotDetails[6]

	if isValidGridSquare( strGrid ) :
		with open( "log.txt", "a" ) as log_file:
			log_file.write( time.ctime() + " : %s is a valid grid square\n" % strGrid )

	else:
		with open("log.txt", "a") as log_file:
			log_file.write( time.ctime() + " : %s is NOT a valid grid square\n" % strGrid )

		sys.exit( 0 )

	latitudeLongitude = gridToLatLong(strGrid).split()

	strLatitude  = toDegreesMinutes( latitudeLongitude[0], 'Latitude' )
	strLongitude = toDegreesMinutes( latitudeLongitude[1], 'Longitude' )

	# save the spot for posterity
	with open( "wspr_spotsLog.txt", "a" ) as spotsLog_file:
		spotsLog_file.write( time.ctime() + " : %s \n" % wsprSpot )

	AprsSendPacket(strUserCallSign, strBalloonCallSign, strLatitude, strLongitude, strIcon, wsprSpot)

else:
	with open( 'log.txt', 'a' ) as log_file:
		log_file.write( time.ctime() + ' : wspr spot is less than 15 minutes older than latest aprs position\n' )
		log_file.close()


# last line for this log entry
with open( "log.txt", "a" ) as log_file:
    log_file.write( time.ctime() + " :#----------------------------------------------------------------------------------------------------------------------------------\n" )

sys.exit( 1 )

# ------------------------------------------------------- That's all Folks!!!! ------------------------------------------------
